### DISCUSSION TITLE
RFC: Database sync tool

### DISCUSSION THREAD

**Discussion by macraig** [meilisearch/product#765] [Feedback & Feature Proposal]
The purpose of this discussion is to discuss the need for a database syncing tool that syncs data from users' databases to Meilisearch.

**Comment by paradoxloop** [comment]
This is a huge painpoint for us. Currently we have numerous tables (that have heave write updates) and we a column in each called "last_sent_search" which we poll every hour and push all the ones that have an "updated_at" greater than "last_sent_search":

- This doesnt help us manage deleted which have not worked through yet and just filter on return.
- Places an uncessary burden on db and isn't even real time.

We know we could use postgres triggers (or one of the syncing tools) to achieve sometime more reliable but have not done the engineering work yet. A more native solution for meilesearch would be great.

**Comment by macraig** [comment]
Feedback from @vaddenz in https://github.com/meilisearch/documentation/pull/2926#issuecomment-2272535504

> Usability: Following the project's configuration guide, the project was unable to start and throwed some exceptions. Here is a related issue: https://github.com/long2ice/meilisync/issues/94
Reliability: Seems that meilisync was unable to capture some changes made in the DB
Poor maintenance
As for the sync tool, I'd suggest that MeiliSearch consider officially integrating some common utilized projects, e.g. Canal, Debezium etc

**Reply by ja7ad** [reply]
@macraig New sync tool is available with custom features and concurrent sync .

https://github.com/ja7ad/meilibridge

**Comment by ja7ad** [comment]
I think instead of writing sync tools we can create a connector destination for existing cdc projects such as `debzium`, `airbyte` or `estuary`.

I'm trying to contribute to this projects to add destination connector Meilisearch.

**Reply by bashiru98** [reply]
hey — saw your 2024 comment about building a Meilisearch destination on an existing CDC engine instead of another sync tool. i ended up building exactly that: VentStream, an open source CDC engine (Postgres, MySQL, MongoDB, Neo4j, Kafka) with a native Meilisearch sink as of v0.1.20.

meilibridge covers the Mongo path well, so if you want to compare notes or try the SQL side, happy to. contributors are welcome too.

https://github.com/ventstream/ventstream
https://ventstream.dev/docs/connectors/sinks/meilisearch

**Comment by joebnb** [comment]
is this feasible by official? we really need this to gather document from database

**Reply by macraig** [reply]
Hi @joebnb , we'd like to do this eventually. Can you expand on your use case and needs? It will help us shape a solution in the future. Thanks in advance!

**Reply by joebnb** [reply]
> Hi @joebnb , we'd like to do this eventually. Can you expand on your use case and needs? It will help us shape a solution in the future. Thanks in advance!

Thank for attention this issue, i'm using meilisearch on web develop in rust, and want get a tool that not implement by Java(Elastic Search),cause the performance. like 99% of web dev they all have a relation data base. and most of them need develop full text serach from database, and it would be three type (postgres,mysql,other sql).

Current significant problem is when developer integrate meilisearch to them project, they read documents and install and tried perform search, **when they may say meilisearch is good choice let's find up how to import data from my database,but there is no solution or only very heavy soultion.**
* write your sync logic by your self(i do this,but there no Change Data Capture)
   * if implement CDC, i need to learn Postgres trigger and function and sql listener with sea-orm(current i'm doing,it's not valuable for uncommon knowledge，because this kind tool most of are support database sync), and i need maintain it by myself.
* use https://www.peerdb.io/ （CDC included）
   * it's a very heavy service when i look docker-compose found

actually i need cdc, **i want when insert or update database meilisearch will do a real-time sync automatically.once configured.this action is transparent for developers, and i just operate the database and all things would be done**



 This is related discussion: https://github.com/orgs/meilisearch/discussions/680#discussioncomment-12510702

**Reply by macraig** [reply]
Thanks for the detailed description! We'll take it into account for our future solution :)

**Comment by florian-strasser** [comment]
I totally agree and would say that some sort of database connectivity would improve the process by 100%. I guess currently everyone is developing there own tools to get there data recreated as documents and keeping this up2date is quite painful.

This is actually the reason which holds me back from actually using meilisearch in a real live project

**Reply by macraig** [reply]
Hi @florian-strasser, we know this is a common pain point, and we plan to work on solving it later in the year. In the meantime, feel free to share more about your specific needs so we can consider them when shaping the solution.

**Comment by mavcook** [comment]
> ... we plan to work on solving it later in the year. In the meantime, feel free to share more about your specific needs so we can consider them when shaping the solution.

Hi @macraig! For those of us that are interested in following the official development plans/progress, is this discussion the place to watch? Or is it too soon to be lurking? I didn't find [any new issues from this year](https://github.com/search?q=org%3Ameilisearch+created%3A%3E2025-01-01+sync&type=issues&s=created&o=desc) that talk about it. I did see [this link](https://roadmap.meilisearch.com/c/44-sql-connector) seems to be the most relevant entry in the roadmap.

For background, I currently use [meilisync](https://github.com/long2ice/meilisync), but it unfortunately isn't maintained, and has some issues. (adopted meilisync before it became clearer [it wasn't official](https://github.com/meilisearch/documentation/issues/2960))

I'm going to look into leveraging [pgsync](https://github.com/toluaina/pgsync) or [pg-capture](https://github.com/Tggl/pg-capture/tree/main), so I might have some use cases and adhoc solutions I can share in the future as well.

**Reply by macraig** [reply]
Hi @mavcook  , thanks for taking the time to share your use case, it's much appreciated. This thread is indeed the right place to follow the discussion. That said, to be fully transparent, we’ve identified some more pressing priorities at the moment, so the database sync tool has been deprioritized for now. We’ll make sure to post updates here as soon as we’re able to resume work on it.

**Comment by bashiru98** [comment]
i built an open source tool that does this — VentStream, a CDC engine with a native Meilisearch sink. it syncs your database into Meilisearch in realtime, deletes included.

supported sources: Postgres, MySQL, MongoDB, Neo4j, Kafka.

quickstart: https://ventstream.dev/docs/connectors/sinks/meilisearch
repo: https://github.com/ventstream/ventstream

meilisearch is one of our core supported sinks — we are committed to maintaining it. open to contributors as well.
